File: gpt-4-turbo/stage2/acc_memcpy_to_device.f90
MODULE acc_test_mod
  USE ISO_C_BINDING
  USE OPENACC
  IMPLICIT NONE
  CONTAINS
!T1:acc memcpy to device,V:2.7-2.3
  LOGICAL FUNCTION test1()
    INTEGER, PARAMETER :: N = 10
    REAL, DIMENSION(N) :: hostArray
    REAL, DIMENSION(:), POINTER :: deviceArray
    INTEGER(C_SIZE_T) :: bytes
    INTEGER :: i, errors

    ! Initialize test conditions
    errors = 0

    ! Initialize hostArray
    DO i = 1, N
      hostArray(i) = REAL(i)
    END DO

    ! Allocate memory on the device
    bytes = N * SIZEOF(hostArray(1))
    deviceArray = C_F_POINTER(acc_malloc(bytes), SHAPE=[N])

    ! Copy data from host to device
    CALL acc_memcpy_to_device(deviceArray, hostArray, bytes)

    ! Verify the copy using a device operation (pseudo code, relies on compiler/runtime support)
    !$acc parallel loop copyin(deviceArray(:N)) reduction(+:errors)
    DO i = 1, N
      IF (deviceArray(i) /= REAL(i)) THEN
        errors = errors + 1
      END IF
    END DO
    !$acc end parallel loop

    ! Cleanup
    CALL acc_free(C_LOC(deviceArray))

    ! Set return value based on errors count
    IF (errors == 0) THEN
      test1 = .FALSE.
    ELSE
      test1 = .TRUE.
    END IF
  END FUNCTION test1
END MODULE acc_test_mod

PROGRAM main
  USE acc_test_mod
  IMPLICIT NONE
  INTEGER :: failcode, testrun
  LOGICAL :: failed
  LOGICAL :: test1

  failed = .FALSE.
  failcode = 0

  DO testrun = 1, NUM_TEST_CALLS
    failed = failed .or. test1()
  END DO
  IF (failed) THEN
    failcode = failcode + 2 ** 0
    failed = .FALSE.
  END IF

  CALL EXIT(failcode)
END PROGRAM main

File: gpt-4-turbo/stage2/acc_memcpy_to_device.cpp
#include <iostream>
#include <stdlib.h>
#include <openacc.h>

#define SEED 12345
#define NUM_TEST_CALLS 10

#ifndef T1
// T1:acc_memcpy_to_device, V:2.7-3.3
int test1() {
    int err = 0;
    srand(SEED);

    size_t num_elements = 1024;    // Number of elements to transfer
    size_t bytes = num_elements * sizeof(int); // Size in bytes

    // Allocate memory on the host
    int *data_host_src = (int *)malloc(bytes);
    if (data_host_src == NULL) {
        std::cerr << "Failed to allocate host memory" << std::endl;
        return 1;
    }

    // Initialize host memory
    for (size_t i = 0; i < num_elements; ++i) {
        data_host_src[i] = rand() % 100; // Random values
    }

    // Allocate memory on the device using OpenACC
    int *data_dev_dest = (int *)acc_malloc(bytes);
    if (data_dev_dest == NULL) {
        std::cerr << "Failed to allocate device memory" << std::endl;
        free(data_host_src);
        return 1;
    }

    // Use acc_memcpy_to_device to transfer data from host to device
    acc_memcpy_to_device(data_dev_dest, data_host_src, bytes);

    // Allocate memory on host for verification
    int *data_verify = (int *)malloc(bytes);
    if (data_verify == NULL) {
        std::cerr << "Failed to allocate verification memory" << std::endl;
        acc_free(data_dev_dest);
        free(data_host_src);
        return 1;
    }

    // Copy data back to host for verification
    acc_memcpy_from_device(data_verify, data_dev_dest, bytes);

    // Verify data
    for (size_t i = 0; i < num_elements; ++i) {
        if (data_host_src[i] != data_verify[i]) {
            err = 1; // Error if the data doesn't match
            break;
        }
    }

    // Cleanup
    free(data_host_src);
    free(data_verify);
    acc_free(data_dev_dest);

    return err;
}
#endif

int main() {
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x) {
        failed = failed + test1();
    }
    if (failed != 0) {
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}